File: src/benchmark.rs
use anyhow::Result;
use openvino::Tensor;
use serde::Serialize;
use std::time::{Duration, Instant};

use crate::engine::Engine;

#[derive(Debug, Clone, Serialize)]
pub struct BenchmarkConfig {
    pub warmup_iters: usize,
    pub duration_secs: f64,
    pub max_iters: Option<usize>,
    pub report_every_secs: f64,
}

#[derive(Debug, Clone, Serialize)]
pub struct LatencyStats {
    pub min_ms: f64,
    pub mean_ms: f64,
    pub p50_ms: f64,
    pub p90_ms: f64,
    pub p95_ms: f64,
    pub p99_ms: f64,
    pub max_ms: f64,
}

#[derive(Debug, Clone, Serialize)]
pub struct BenchmarkReport {
    pub warmup_iters: usize,
    pub measured_iters: usize,
    pub measured_seconds: f64,
    pub throughput_fps: f64,
    pub latency: LatencyStats,
}

#[derive(Debug, Clone, Serialize)]
pub struct StageTimingReport {
    pub iterations: usize,
    pub stage_read_each_iter: bool,
    pub io_decode_mean_ms: f64,
    pub preprocess_mean_ms: f64,
    pub inference_mean_ms: f64,
    pub postprocess_mean_ms: f64,
    pub total_mean_ms: f64,
    pub io_decode_share_pct: f64,
    pub preprocess_share_pct: f64,
    pub inference_share_pct: f64,
    pub postprocess_share_pct: f64,
}

pub fn run_benchmark(
    engine: &mut Engine,
    tensor: &Tensor,
    cfg: &BenchmarkConfig,
) -> Result<BenchmarkReport> {
    let mut request = engine.create_request()?;

    run_benchmark_loop(cfg, || engine.run_request(&mut request, tensor))
}

pub fn run_benchmark_loop<F>(cfg: &BenchmarkConfig, mut step: F) -> Result<BenchmarkReport>
where
    F: FnMut() -> Result<()>,
{
    for _ in 0..cfg.warmup_iters {
        step()?;
    }

    let mut latencies_us: Vec<u64> = Vec::new();
    let start = Instant::now();
    let mut last_report = start;
    let mut measured_iters = 0usize;

    loop {
        if let Some(max_iters) = cfg.max_iters {
            if measured_iters >= max_iters {
                break;
            }
        } else if start.elapsed().as_secs_f64() >= cfg.duration_secs {
            break;
        }

        let iter_start = Instant::now();
        step()?;
        let elapsed = iter_start.elapsed();
        latencies_us.push(duration_to_micros(elapsed));
        measured_iters += 1;

        if cfg.report_every_secs > 0.0
            && last_report.elapsed().as_secs_f64() >= cfg.report_every_secs
        {
            let secs = start.elapsed().as_secs_f64();
            let fps = measured_iters as f64 / secs.max(1e-9);
            eprintln!("Benchmark progress: {measured_iters} iters, {fps:.2} img/s");
            last_report = Instant::now();
        }
    }

    let measured_seconds = start.elapsed().as_secs_f64();
    let throughput_fps = measured_iters as f64 / measured_seconds.max(1e-9);
    let latency = compute_latency_stats(&latencies_us);

    Ok(BenchmarkReport {
        warmup_iters: cfg.warmup_iters,
        measured_iters,
        measured_seconds,
        throughput_fps,
        latency,
    })
}

fn compute_latency_stats(latencies_us: &[u64]) -> LatencyStats {
    if latencies_us.is_empty() {
        return LatencyStats {
            min_ms: 0.0,
            mean_ms: 0.0,
            p50_ms: 0.0,
            p90_ms: 0.0,
            p95_ms: 0.0,
            p99_ms: 0.0,
            max_ms: 0.0,
        };
    }

    let mut sorted = latencies_us.to_vec();
    sorted.sort_unstable();

    let sum_us: u128 = latencies_us.iter().map(|&x| x as u128).sum();
    let mean_us = sum_us as f64 / latencies_us.len() as f64;

    LatencyStats {
        min_ms: us_to_ms(*sorted.first().unwrap_or(&0)),
        mean_ms: us_to_ms_f64(mean_us),
        p50_ms: percentile_ms(&sorted, 50.0),
        p90_ms: percentile_ms(&sorted, 90.0),
        p95_ms: percentile_ms(&sorted, 95.0),
        p99_ms: percentile_ms(&sorted, 99.0),
        max_ms: us_to_ms(*sorted.last().unwrap_or(&0)),
    }
}

fn percentile_ms(sorted_us: &[u64], p: f64) -> f64 {
    if sorted_us.is_empty() {
        return 0.0;
    }
    let n = sorted_us.len();
    let rank = ((p / 100.0) * (n as f64 - 1.0)).round() as usize;
    us_to_ms(sorted_us[rank.min(n - 1)])
}

fn duration_to_micros(d: Duration) -> u64 {
    d.as_micros().min(u64::MAX as u128) as u64
}

fn us_to_ms(us: u64) -> f64 {
    us as f64 / 1000.0
}

fn us_to_ms_f64(us: f64) -> f64 {
    us / 1000.0
}

File: python/inference_rs/benchmark.py
"""Benchmark loop and latency statistics.

Matches the Rust ``benchmark.rs`` implementation:
- Warmup N iterations (results discarded)
- Measurement loop: time-based or iteration-limited
- Microsecond-resolution latency measurements
- Nearest-rank percentile: ``rank = round((P/100) * (n-1))``
- Throughput: ``measured_iters / max(measured_seconds, 1e-9)``
"""

from __future__ import annotations

import sys
import time
from dataclasses import dataclass, field
from typing import Callable


@dataclass
class BenchmarkConfig:
    warmup_iters: int = 20
    duration_secs: float = 10.0
    max_iters: int | None = None
    report_every_secs: float = 1.0

    def to_dict(self) -> dict:
        return {
            "warmup_iters": self.warmup_iters,
            "duration_secs": self.duration_secs,
            "max_iters": self.max_iters,
            "report_every_secs": self.report_every_secs,
        }


@dataclass
class LatencyStats:
    min_ms: float = 0.0
    mean_ms: float = 0.0
    p50_ms: float = 0.0
    p90_ms: float = 0.0
    p95_ms: float = 0.0
    p99_ms: float = 0.0
    max_ms: float = 0.0

    def to_dict(self) -> dict:
        return {
            "min_ms": self.min_ms,
            "mean_ms": self.mean_ms,
            "p50_ms": self.p50_ms,
            "p90_ms": self.p90_ms,
            "p95_ms": self.p95_ms,
            "p99_ms": self.p99_ms,
            "max_ms": self.max_ms,
        }


@dataclass
class BenchmarkReport:
    warmup_iters: int = 0
    measured_iters: int = 0
    measured_seconds: float = 0.0
    throughput_fps: float = 0.0
    latency: LatencyStats = field(default_factory=LatencyStats)

    def to_dict(self) -> dict:
        return {
            "warmup_iters": self.warmup_iters,
            "measured_iters": self.measured_iters,
            "measured_seconds": self.measured_seconds,
            "throughput_fps": self.throughput_fps,
            "latency": self.latency.to_dict(),
        }


@dataclass
class StageTimingReport:
    iterations: int = 0
    stage_read_each_iter: bool = False
    io_decode_mean_ms: float = 0.0
    preprocess_mean_ms: float = 0.0
    inference_mean_ms: float = 0.0
    postprocess_mean_ms: float = 0.0
    total_mean_ms: float = 0.0
    io_decode_share_pct: float = 0.0
    preprocess_share_pct: float = 0.0
    inference_share_pct: float = 0.0
    postprocess_share_pct: float = 0.0

    def to_dict(self) -> dict:
        return {
            "iterations": self.iterations,
            "stage_read_each_iter": self.stage_read_each_iter,
            "io_decode_mean_ms": self.io_decode_mean_ms,
            "preprocess_mean_ms": self.preprocess_mean_ms,
            "inference_mean_ms": self.inference_mean_ms,
            "postprocess_mean_ms": self.postprocess_mean_ms,
            "total_mean_ms": self.total_mean_ms,
            "io_decode_share_pct": self.io_decode_share_pct,
            "preprocess_share_pct": self.preprocess_share_pct,
            "inference_share_pct": self.inference_share_pct,
            "postprocess_share_pct": self.postprocess_share_pct,
        }


def run_benchmark_loop(
    cfg: BenchmarkConfig, step: Callable[[], None]
) -> BenchmarkReport:
    """Run the benchmark loop: warmup, then timed measurement.

    *step* is called once per iteration and should perform one complete
    inference pass.
    """
    # Warmup
    for _ in range(cfg.warmup_iters):
        step()

    latencies_us: list[int] = []
    start = time.perf_counter()
    last_report = start
    measured_iters = 0

    while True:
        if cfg.max_iters is not None:
            if measured_iters >= cfg.max_iters:
                break
        elif time.perf_counter() - start >= cfg.duration_secs:
            break

        iter_start = time.perf_counter()
        step()
        elapsed = time.perf_counter() - iter_start
        latencies_us.append(int(elapsed * 1_000_000))
        measured_iters += 1

        if cfg.report_every_secs > 0.0:
            now = time.perf_counter()
            if now - last_report >= cfg.report_every_secs:
                secs = now - start
                fps = measured_iters / max(secs, 1e-9)
                print(
                    f"Benchmark progress: {measured_iters} iters, {fps:.2f} img/s",
                    file=sys.stderr,
                )
                last_report = now

    measured_seconds = time.perf_counter() - start
    throughput_fps = measured_iters / max(measured_seconds, 1e-9)
    latency = _compute_latency_stats(latencies_us)

    return BenchmarkReport(
        warmup_iters=cfg.warmup_iters,
        measured_iters=measured_iters,
        measured_seconds=measured_seconds,
        throughput_fps=throughput_fps,
        latency=latency,
    )


def _compute_latency_stats(latencies_us: list[int]) -> LatencyStats:
    if not latencies_us:
        return LatencyStats()

    sorted_us = sorted(latencies_us)
    sum_us = sum(latencies_us)
    mean_us = sum_us / len(latencies_us)

    return LatencyStats(
        min_ms=sorted_us[0] / 1000.0,
        mean_ms=mean_us / 1000.0,
        p50_ms=_percentile_ms(sorted_us, 50.0),
        p90_ms=_percentile_ms(sorted_us, 90.0),
        p95_ms=_percentile_ms(sorted_us, 95.0),
        p99_ms=_percentile_ms(sorted_us, 99.0),
        max_ms=sorted_us[-1] / 1000.0,
    )


def _percentile_ms(sorted_us: list[int], p: float) -> float:
    """Nearest-rank percentile matching Rust: ``rank = round((P/100) * (n-1))``."""
    if not sorted_us:
        return 0.0
    n = len(sorted_us)
    rank = round((p / 100.0) * (n - 1))
    rank = min(rank, n - 1)
    return sorted_us[rank] / 1000.0
